sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c

```diff
@@ -299,19 +280,24 @@ void gnrc_lorawan_process_fopts(gnrc_lorawan_t *mac, uint8_t *fopts, size_t size
         return;
     }
 
-    lorawan_buffer_t buf;
-    _buffer_reset(&buf, fopts, size);
+    uint8_t ret = 0;
+    int (*cb)(gnrc_lorawan_t*, uint8_t *p) = NULL;
 
-    while (buf.index < buf.size) {
-        switch (*(buf.data)) {
+    for(uint8_t pos = 0; pos < size; pos += ret) {
+        switch (fopts[pos]) {
             case GNRC_LORAWAN_CID_LINK_CHECK_REQ_ANS:
-                if (_mlme_link_check_ans(mac, &buf) < 0) {
-                    return;
-                }
+                ret += 3;
+                cb = _mlme_link_check_ans;
                 break;
             default:
                 return;
         }
+
+        if(pos + ret > size) {
+            return;
+        }
```

Comment:
IMHO this is redundant with the for-loop condition

Reply:
the problem is, I have to check that it doesn't overflow before executing the fopt payload

Reply:
it's a tri-state (`pos + ret < size`, `pos + rest == size`, `pos + ret > size`). Each one of them has a different behavior.

Reply:
ah so this should work then:

```
for(uint8_t pos = 0; pos < (size - 2); pos += ret) {
    switch (fopts[pos]) {
    case GNRC_LORAWAN_CID_LINK_CHECK_REQ_ANS:
        ret += 3;
        _mlme_link_check_ans(mac, &fopts[pos + 1]);
        break;
    default:
        return;
    }
}
```

please note: I changed the pointer to be `pos + 1` because you access `p[0]` and `p[1]` in `_mlme_link_check_ans`.

Reply:
That loop doesn't work :(

If size is e.g 2 (the minimum fopt size is 2), then the loop would never be executed

Reply:
done! (https://github.com/RIOT-OS/RIOT/pull/11022/commits/22aa1aee7f670a164f742704f801cd1df6e6ab4c)
```
2019-11-22 13:40:21,070 # Iface  4  HWaddr: 26:01:2B:7D  Frequency: 869524963Hz  BW: 125kHz  SF: 9  CR: 4/5  Link: up 
2019-11-22 13:40:21,077 #            TX-Power: 14dBm  State: SLEEP  Demod margin.: 24  Num gateways.: 2 
2019-11-22 13:40:21,080 #           IQ_INVERT 
2019-11-22 13:40:21,083 #           RX_SINGLE OTAA L2-PDU:255 
```